src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal sealed class OpenSslPkcs12Reader : IDisposable
+    {
+        private readonly SafePkcs12Handle _pkcs12Handle;
+        private SafeEvpPkeyHandle _evpPkeyHandle;
+        private SafeX509Handle _x509Handle;
+        private SafeX509StackHandle _caStackHandle;
+
+        private OpenSslPkcs12Reader(SafePkcs12Handle pkcs12Handle)
+        {
+            _pkcs12Handle = pkcs12Handle;
+        }
+
+        public unsafe static OpenSslPkcs12Reader TryRead(byte[] data)
+        {
+            SafePkcs12Handle handle = Interop.libcrypto.OpenSslD2I(
+                (ptr, b, i) => Interop.libcrypto.d2i_PKCS12(ptr, b, i),
+                data,
+                checkHandle: false);
+
+            if (handle != null && !handle.IsInvalid)
+            {
+                return new OpenSslPkcs12Reader(handle);
+            }
+
+            return null;
+        }
+
+        public static OpenSslPkcs12Reader TryRead(SafeBioHandle fileBio)
```

Comment:
I was wondering the same thing.  Rather than the usage pattern currently employed:
```C#
using (OpenSslPkcs12Reader reader = TryRead(handle))
{
    if (reader != null)
    {
        ...
    }
}
```
I wonder whether it'd be cleaner / more familiar to have the reader be an out parameter and do:
```C#
OpenSslPkcs12Reader reader;
if (TryRead(handle, out reader))
{
    using (reader)
    {
        ...
    }
}
```
or something like that.  I'm ok with the current code though if it ends up working best.